domain/cloudimagemetadata/state/state.go

```diff
@@ -0,0 +1,344 @@
+//
+//   - Stream
+//   - Region
+//   - Version
+//   - Arch
+//   - VirtType
+//   - RootStorageType
+//   - Source
+//
+// Above behaviors applies for duplicated inserted metadata:
+//   - If a metadata has the same signature than an existing one in the database, the imageID will be
+//     updated with the new value in the existing metadata
+//   - If a several metadatas have the same signature in the list, the first one will be inserted, then
+//     the imageID of this metadata will be updated to the last one (exactly as if they were saved in
+//     several consecutive calls.
+func (s *State) SaveMetadata(ctx context.Context, metadatas []cloudimagemetadata.Metadata) error {
+	// Cleaning the database before insert/update is a reasonable way to remove outdated data without
+	// requiring an housekeeping goroutine. Since a metadata data doesn't live long in the db and
+	// it shouldn't be a lot of custom metadata, it shouldn't slow too much the insertion.
+	// Expired metadata are ignored when retrieving metadata, so it doesn't have any impact to keep
+	// some of them longer than their time to live.
+	if err := s.tryCleanUpExpiredMetadata(ctx); err != nil {
+		return errors.Trace(err)
+	}
+	if err := s.ensuresArchitecturesLoaded(ctx); err != nil {
+		return errors.Trace(err)
+	}
+
+	db, err := s.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// Prepare inputs
+	createdAt := time.Now()
+	values := make([]inputMetadata, 0, len(metadatas))
+	for _, m := range metadatas {
+		err := validateMetadata(m)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		// Convert architecture name to a db id
+		archId, ok := s.architectureIDsByName[m.Arch]
+		if !ok {
+			s.logger.Warningf("skipping metadata for unknown architecture %q", m.Arch)
+			continue
+		}
+
+		// Update creation time if not defined.
+		var noTime time.Time
+		if m.CreationTime != noTime {
+			createdAt = m.CreationTime
+		}
+
+		mUUID, err := uuid.NewUUID()
+		if err != nil {
+			return errors.Annotatef(err, "failed to generate metadata uuid")
+		}
+		values = append(values, inputMetadata{
+			UUID:            mUUID.String(),
+			CreatedAt:       createdAt,
+			Source:          m.Source,
+			Stream:          m.Stream,
+			Region:          m.Region,
+			Version:         m.Version,
+			VirtType:        m.VirtType,
+			ArchitectureID:  archId,
+			RootStorageType: m.RootStorageType,
+			RootStorageSize: m.RootStorageSize,
+			Priority:        m.Priority,
+			ImageID:         m.ImageId,
+		})
+	}
+
+	// Prepare statement to insert a batch of image in db. In case of conflict on unique constraint,
+	// it will update the image id instead
+	insertMetadataStmt, err := sqlair.Prepare(`
+INSERT INTO cloud_image_metadata (*)
+VALUES ($inputMetadata.*)
+ON CONFLICT (stream, region, version, architecture_id, virt_type, root_storage_type, source) DO UPDATE
+SET image_id = EXCLUDED.image_id;`, inputMetadata{})
```

Comment:
So this should be changed. We're steering clear of ON CONFLICT because it's harder to reason about. So instead we should use a select and then based on that, insert or update if required. This gives us also the advantage of not moving to a write transaction, whilst a read could prevent additional writes.

Reply:
Are you 100% sure ? I see this pattern a few time in quite recent code (2024), it's why I don't discard it by default (I am aware teams tend to avoid it, mostly because it is non standard SQL).

Removing this clause will make the save code a lot more complex, whereas actual request is straightforward (on conflict, just update the image ID). It's even more readable than in the mongodb version.

A way to keep it simple to read but may be less efficient would be to remove the batch insert from state layer (and not push a slice into `SaveMetadata`, but a single value, and loop on service layer). This compromise, however, has drawback, since it will fire a lot mot the cleanup, unless we implements a clever cleanup strategy (which implies... strange code with semaphore, so we mostly shift complexity on read), and generate a hell more transaction (one by metadata)

I have chosen the `INSERT ON CONFLICT` has the better compromise between readability and resources saving.

What is your final cut ? ;)

Reply:
@SimonRichardson I have removed the upsert in commit https://github.com/juju/juju/pull/18082/commits/d18a2de67b0f92ec7dfd634706332492326356d7

However, I still don't find it is a good solution, and @nvinuesa doesn't seems to have any issue regarding upsert in this case.

I have pushed the review fix commit by commit to make it possible to rollback this one, if you change your mind.